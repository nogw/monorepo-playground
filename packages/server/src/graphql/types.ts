import { Maybe } from '@fp/types';

import { DataLoaders } from '../modules/loader/loaderRegister';
import { IUserDocument } from '../modules/user/UserModel';

export type GraphQLContext = {
  dataloaders: DataLoaders;
  user?: Maybe<IUserDocument>;
};
