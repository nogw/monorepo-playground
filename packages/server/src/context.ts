import { Request } from 'koa';

import { Maybe } from '@fp/types';

import { getDataloaders } from './modules/loader/loaderRegister';
import { GraphQLContext } from './graphql/types';
import { IUserDocument } from './modules/user/UserModel';

type ContextVariables = {
  user?: Maybe<IUserDocument>;
  req?: Request;
};

export const getContext = (context?: ContextVariables): GraphQLContext => {
  const dataloaders = getDataloaders();

  return {
    dataloaders,
    user: context?.user,
  };
};
