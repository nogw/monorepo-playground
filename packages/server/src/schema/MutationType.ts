import { GraphQLObjectType } from 'graphql';

import * as UserMutation from '../modules/user/mutations';

export const MutationType = new GraphQLObjectType({
  name: 'Mutation',
  description: 'Root of all mutations',
  fields: () => ({
    ...UserMutation,
  }),
});
