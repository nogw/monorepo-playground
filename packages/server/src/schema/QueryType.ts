import { GraphQLObjectType } from 'graphql';

import { nodeField, nodesField } from '../modules/node/typeRegister';
import UserLoader from '../modules/user/UserLoader';
import UserType from '../modules/user/UserType';

export const QueryType = new GraphQLObjectType({
  name: 'Query',
  description: 'Root of all queries',
  fields: () => ({
    nodes: nodesField,
    node: nodeField,
    me: {
      type: UserType,
      resolve: (_root, _args, context) => {
        return UserLoader.load(context, context.user?.id);
      },
    },
  }),
});
