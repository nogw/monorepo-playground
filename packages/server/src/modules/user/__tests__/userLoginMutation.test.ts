import { graphql } from 'graphql';

import {
  clearDatabaseAndRestartCounters,
  connectWithMongoose,
  disconnectWithMongoose,
  sanitizeTestObject,
} from '../../../../test';
import { createUser } from '../fixtures/createUser';
import { getContext } from '../../../context';
import { schema } from '../../../schema/schema';

beforeAll(connectWithMongoose);
beforeEach(clearDatabaseAndRestartCounters);
afterAll(disconnectWithMongoose);

describe('UserLoginMutation', () => {
  it('should login with a registered user', async () => {
    const { username, email } = await createUser({
      email: 'nogw@nogw.com',
      password: 'a9218c490c89864790bf',
    });

    const mutation = `
      mutation UserLoginMutation($input: UserLoginInput!) {
        userLogin(input: $input) {
          error
          token
          me {
            username
          }
        }
      }
    `;

    const rootValue = {};
    const contextValue = getContext();
    const variables = {
      email: email,
      password: 'a9218c490c89864790bf',
    };

    const result = await graphql({
      schema,
      rootValue,
      contextValue,
      source: mutation,
      variableValues: { input: variables },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data.userLogin.error).toBeNull();
    expect(result.data.userLogin.token).toBeDefined();
    expect(result.data.userLogin.me.username).toBe(username);
    expect(sanitizeTestObject(result.data, ['token'])).toMatchSnapshot();
  });

  it("should display error if username isn't exists", async () => {
    const mutation = `
      mutation UserLoginMutation($input: UserLoginInput!) {
        userLogin(input: $input) {
          error
          token
          me {
            id
          }
        }
      }
    `;

    const rootValue = {};
    const contextValue = getContext();
    const variables = {
      email: 'awesome@fpmail.com',
      password: 'awesomepassword',
    };

    const result = await graphql({
      schema,
      rootValue,
      contextValue,
      source: mutation,
      variableValues: { input: variables },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data.userLogin.token).toBeNull();
    expect(result.data.userLogin.me).toBeNull();
    expect(result.data.userLogin.error).toBe('This user was not registered');
  });

  it('should display error if password is incorrect', async () => {
    const { email } = await createUser({ email: 'awesome@fpmail.com' });

    const mutation = `
      mutation UserLoginMutation($input: UserLoginInput!) {
        userLogin(input: $input) {
          error
          token
          me {
            id
          }
        }
      }
    `;

    const rootValue = {};
    const contextValue = getContext();
    const variables = {
      email: email,
      password: 'awesomepassword',
    };

    const result = await graphql({
      schema,
      rootValue,
      contextValue,
      source: mutation,
      variableValues: { input: variables },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data.userLogin.token).toBeNull();
    expect(result.data.userLogin.me).toBeNull();
    expect(result.data.userLogin.error).toBe('This password is incorrect');
  });
});
