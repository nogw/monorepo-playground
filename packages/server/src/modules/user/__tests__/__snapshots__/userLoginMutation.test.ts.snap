// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserLoginMutation should login with a registered user 1`] = `
{
  "userLogin": {
    "error": "EMPTY",
    "me": {
      "username": "user#0",
    },
    "token": "FROZEN-TOKEN",
  },
}
`;
